File: src/main/intellibitz/sted/io/FileFilterHelper.java
package intellibitz.sted.io;

import javax.swing.filechooser.FileFilter;
import java.io.File;
import java.util.HashMap;
import java.util.Hashtable;
import java.util.Iterator;
import java.util.Map;

/**
 * A convenience implementation of FileFilter that filters out all files except
 * for those type extensions that it knows about.
 * <p/>
 * Extensions are of the type ".foo", which is typically found on Windows and
 * Unix boxes, but not on Macinthosh. Case is ignored.
 * <p/>
 * Example - create a new filter that filerts out all files but gif and jpg
 * image files:
 * <p/>
 * JFileChooser chooser = new JFileChooser(); SideFileFilter filter = new
 * SideFileFilter( new String{"gif", "jpg"}, "JPEG & GIF Images")
 * chooser.addChoosableFileFilter(filter); chooser.showOpenDialog(this);
 */
public class FileFilterHelper
        extends FileFilter {

//    private static String TYPE_UNKNOWN = "Type Unknown";
//    private static String HIDDEN_FILE = "Hidden File";

    private Map<String, FileFilterHelper> filters;
    private String description;
    private String fullDescription;
    private boolean useExtensionsInDescription = true;

    /**
     * Creates a file filter. If no filters are added, then all files are
     * accepted.
     *
     * @see #addExtension
     */
    public FileFilterHelper() {
        filters = new HashMap<String, FileFilterHelper>();
    }

    /**
     * Creates a file filter that accepts files with the given extension.
     * Example: new FileFilterHelper("jpg");
     *
     * @see #addExtension
     */
    public FileFilterHelper(String extension) {
        this(extension, null);
    }

    /**
     * Creates a file filter that accepts the given file type. Example: new
     * FileFilterHelper("jpg", "JPEG Image Images");
     * <p/>
     * Note that the "." before the extension is not needed. If provided, it
     * will be ignored.
     *
     * @see #addExtension
     */
    public FileFilterHelper(String extension, String description) {
        this();
        if (extension != null) {
            addExtension(extension);
        }
        if (description != null) {
            setDescription(description);
        }
    }

    /**
     * Creates a file filter from the given string array. Example: new
     * FileFilterHelper(String {"gif", "jpg"});
     * <p/>
     * Note that the "." before the extension is not needed adn will be
     * ignored.
     *
     * @see #addExtension
     */
    public FileFilterHelper(String[] filters) {
        this(filters, null);
    }

    /**
     * Creates a file filter from the given string array and description.
     * Example: new FileFilterHelper(String {"gif", "jpg"}, "Gif and JPG
     * Images");
     * <p/>
     * Note that the "." before the extension is not needed and will be
     * ignored.
     *
     * @see #addExtension
     */
    public FileFilterHelper(String[] filters, String description) {
        this();
        for (final String newVar : filters) {
            // add filters one by one
            addExtension(newVar);
        }
        if (description != null) {
            setDescription(description);
        }
    }

    /**
     * Return true if this file should be shown in the directory pane, false if
     * it shouldn't.
     * <p/>
     * Files that begin with "." are ignored.
     *
     * @see #getExtension
     * @see javax.swing.filechooser.FileFilter#accept(java.io.File)
     */
    public boolean accept(File f) {
        if (f != null) {
            if (f.isDirectory()) {
                return true;
            }
            final String extension = getExtension(f);
            if (extension != null && filters.get(getExtension(f)) != null) {
                return true;
            }
        }
        return false;
    }

    /**
     * Return the extension portion of the file's name .
     *
     * @see #getExtension
     * @see FileFilter#accept
     */
    public static String getExtension(File f) {
        if (f != null) {
            final String filename = f.getName();
            final int i = filename.lastIndexOf('.');
            if (i > 0 && i < filename.length() - 1) {
                return filename.substring(i + 1).toLowerCase();
            }
        }
        return null;
    }

    /**
     * Adds a filetype "dot" extension to filter against.
     * <p/>
     * For example: the following code will create a filter that filters out all
     * files except those that end in ".jpg" and ".tif":
     * <p/>
     * FileFilterHelper filter = new FileFilterHelper();
     * filter.addExtension("jpg"); filter.addExtension("tif");
     * <p/>
     * Note that the "." before the extension is not needed and will be
     * ignored.
     */
    public void addExtension(String extension) {
        if (filters == null) {
            filters = new Hashtable<String, FileFilterHelper>(5);
        }
        filters.put(extension.toLowerCase(), this);
        fullDescription = null;
    }


    /**
     * Returns the human readable description of this filter. For example: "JPEG
     * and GIF Image Files (*.jpg, *.png)"
     *
     * @see FileFilter#getDescription
     */
    public String getDescription() {
        if (fullDescription == null) {
            if (description == null || isExtensionListInDescription()) {
                fullDescription =
                        description == null ? "(" : description + " (";
                // build the description from the extension list
                final Iterator<String> extensions = filters.keySet().iterator();
                if (extensions != null) {
                    fullDescription += "." + extensions.next();
                    while (extensions.hasNext()) {
                        fullDescription += ", " + extensions.next();
                    }
                }
                fullDescription += ")";
            } else {
                fullDescription = description;
            }
        }
        return fullDescription;
    }

    /**
     * Sets the human readable description of this filter. For example:
     * filter.setDescription("Gif and JPG Images");
     */
    public void setDescription(String description) {
        this.description = description;
        fullDescription = null;
    }

    /**
     * Determines whether the extension list (.jpg, .png, etc) should show up in
     * the human readable description.
     * <p/>
     * Only relevent if a description was provided in the constructor or using
     * setDescription();
     */
    public void setExtensionListInDescription(boolean b) {
        useExtensionsInDescription = b;
        fullDescription = null;
    }

    /**
     * Returns whether the extension list (.jpg, .png, etc) should show up in
     * the human readable description.
     * <p/>
     * Only relevent if a description was provided in the constructor or using
     * setDescription();
     */
    public boolean isExtensionListInDescription() {
        return useExtensionsInDescription;
    }
}

File: src/main/intellibitz/sted/io/FileFilterHelper.kt
package sted.io

import java.io.File
import java.util.*
import javax.swing.filechooser.FileFilter

/**
 * A convenience implementation of FileFilter that filters out all files except
 * for those type extensions that it knows about.
 * Extensions are of the type ".foo", which is typically found on Windows and
 * Unix boxes, but not on Macinthosh. Case is ignored.
 * Example - create a new filter that filerts out all files but gif and jpg
 * image files:
 * JFileChooser chooser = new JFileChooser(); FileFilterHelper filter = new
 * FileFilterHelper( new String{"gif", "jpg"}, "JPEG & GIF Images")
 * chooser.addChoosableFileFilter(filter); chooser.showOpenDialog(this);
 */
class FileFilterHelper() : FileFilter() {
    private var filters: MutableMap<String?, FileFilterHelper?>?
    private var description: String? = null
    private var fullDescription: String? = null

    /**
     * Creates a file filter that accepts the given file type. Example: new
     * FileFilterHelper("jpg", "JPEG Image Images");
     * Note that the "." before the extension is not needed. If provided, it
     * will be ignored.
     *
     * @see .addExtension
     */
    constructor(extension: String?, description: String?) : this() {
        extension?.let { addExtension(it) }
        description?.let { setDescription(it) }
    }

    /**
     * Return true if this file should be shown in the directory pane, false if
     * it shouldn't.
     * Files that begin with "." are ignored.
     *
     * @see .getExtension
     *
     * @see javax.swing.filechooser.FileFilter.accept
     */
    override fun accept(f: File): Boolean {
        if (f.isDirectory) {
            return true
        }
        val extension = getExtension(f)
        return extension != null && filters!![getExtension(f)] != null
    }

    /**
     * Adds a filetype "dot" extension to filter against.
     * For example: the following code will create a filter that filters out all
     * files except those that end in ".jpg" and ".tif":
     *
     *
     * FileFilterHelper filter = new FileFilterHelper();
     * filter.addExtension("jpg"); filter.addExtension("tif");
     *
     *
     * Note that the "." before the extension is not needed and will be
     * ignored.
     */
    private fun addExtension(extension: String) {
        if (filters == null) {
            filters = Hashtable(5)
        }
        filters!![extension.toLowerCase()] = this
        fullDescription = null
    }

    /**
     * Returns the human readable description of this filter. For example: "JPEG
     * and GIF Image Files (*.jpg, *.png)"
     *
     * @see FileFilter.getDescription
     */
    override fun getDescription(): String {
        if (fullDescription == null) {
            if (description == null || isExtensionListInDescription) {
                fullDescription = if (description == null) "(" else "$description ("
                // build the description from the extension list
                val extensions: Iterator<String?> = filters!!.keys.iterator()
                fullDescription += "." + extensions.next()
                while (extensions.hasNext()) {
                    fullDescription += ", " + extensions.next()
                }
                fullDescription += ")"
            } else {
                fullDescription = description
            }
        }
        return fullDescription!!
    }

    /**
     * Sets the human readable description of this filter. For example:
     * filter.setDescription("Gif and JPG Images");
     */
    private fun setDescription(description: String?) {
        this.description = description
        fullDescription = null
    }

    /**
     * Returns whether the extension list (.jpg, .png, etc) should show up in
     * the human readable description.
     *
     *
     * Only relevent if a description was provided in the constructor or using
     * setDescription();
     */
    private val isExtensionListInDescription: Boolean
        get() = true

    companion object {
        /**
         * Return the extension portion of the file's name .
         *
         * @see .getExtension
         *
         * @see FileFilter.accept
         */
        fun getExtension(f: File?): String? {
            if (f != null) {
                val filename = f.name
                val i = filename.lastIndexOf('.')
                if (i > 0 && i < filename.length - 1) {
                    return filename.substring(i + 1).toLowerCase()
                }
            }
            return null
        }
    }

    /**
     * Creates a file filter. If no filters are added, then all files are
     * accepted.
     *
     * @see .addExtension
     */
    init {
        filters = HashMap()
    }
}